In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

# Use the kagglehub client library to attach Kaggle resources like competitions, datasets, and models to your session
# Learn more about kagglehub: https://github.com/Kaggle/kagglehub/blob/main/README.md

import kagglehub
# kagglehub.dataset_download('<owner>/<dataset-slug>')

/kaggle/input/datasets/muhammadumer7804/sales-dataset/retail_sales.csv


In [2]:
# STEP 3: Loading the Dataset
# We use pandas library to read our CSV file

import pandas as pd

# Load the dataset from the file path
df = pd.read_csv("/kaggle/input/datasets/muhammadumer7804/sales-dataset/retail_sales.csv")

# Print a message to confirm it loaded successfully
print("Dataset loaded successfully!")

Dataset loaded successfully!


In [3]:
# STEP 4: Exploring the Dataset
# Let's see what our data looks like

# Show the first 5 rows of the dataset
print("First 5 rows of the dataset:")
df.head()

First 5 rows of the dataset:


,Date,Category,Sales,Quantity,Profit,Region
0,1/1/2023,Electronics,1149.014246,11.0,383.664245,North
1,1/1/2023,Clothing,958.520710,7.0,224.054049,East
2,1/1/2023,Home Goods,1473.763845,2.0,466.593090,South
3,1/1/2023,Sports,1230.230419,6.0,123.310460,West
4,1/1/2023,NaN?,828.585950,12.0,88.591355,East


In [4]:
# Show the shape of the dataset
# This tells us: (number of rows, number of columns)
print("Dataset Shape (Rows, Columns):", df.shape)

Dataset Shape (Rows, Columns): (1825, 6)


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1825 entries, 0 to 1824
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   Date      1825 non-null   object 
 1   Category  1821 non-null   object 
 2   Sales     1823 non-null   float64
 3   Quantity  1820 non-null   float64
 4   Profit    1825 non-null   float64
 5   Region    1820 non-null   object 
dtypes: float64(3), object(3)
memory usage: 85.7+ KB


In [6]:
df.shape

(1825, 6)

In [7]:
# STEP 5: Finding Missing Values
# This shows exactly how many missing (empty) values are in each column

missing_values = df.isnull().sum()

print("Missing Values in Each Column:")
print(missing_values)

Missing Values in Each Column:
Date        0
Category    4
Sales       2
Quantity    5
Profit      0
Region      5
dtype: int64


In [8]:
# STEP 6: Fixing Missing Values

# For text columns, fill missing values with the most common value (mode)
df['Category'] = df['Category'].fillna(df['Category'].mode()[0])
df['Region'] = df['Region'].fillna(df['Region'].mode()[0])

# For number columns, fill missing values with the median (middle value)
df['Sales'] = df['Sales'].fillna(df['Sales'].median())
df['Quantity'] = df['Quantity'].fillna(df['Quantity'].median())

print("Missing values have been fixed!")

Missing values have been fixed!


In [9]:
# Let's verify that missing values are gone
print("Missing Values After Fixing:")
print(df.isnull().sum())

Missing Values After Fixing:
Date        0
Category    0
Sales       0
Quantity    0
Profit      0
Region      0
dtype: int64


In [10]:
# STEP 7: Finding Duplicate Rows
# This checks if any row is exactly repeated in the dataset

duplicate_count = df.duplicated().sum()

print("Number of Duplicate Rows Found:", duplicate_count)

Number of Duplicate Rows Found: 0


In [11]:
# STEP 8: Saving the Cleaned Dataset
# This saves our cleaned data as a new CSV file

df.to_csv("cleaned_retail_sales.csv", index=False)

print("Cleaned dataset has been saved successfully!")

Cleaned dataset has been saved successfully!


# Simple Data Cleaning Agent 🤖🧹

## Project Overview
This notebook builds a simple AI agent that automatically cleans a messy retail sales dataset. 
The agent detects missing values, fixes them using appropriate strategies, checks for duplicate 
rows, and saves a clean, ready-to-use version of the dataset.

## Dataset
- **Name:** Retail Sales Dataset
- **Rows:** 1825
- **Columns:** 6 (Date, Category, Sales, Quantity, Profit, Region)

## Steps Performed
1. **Load Dataset** — Imported the CSV file using pandas.
2. **Explore Dataset** — Checked the first few rows, data types, and overall shape.
3. **Detect Missing Values** — Found missing values in 4 columns:
   - Category: 4 missing
   - Sales: 2 missing
   - Quantity: 5 missing
   - Region: 5 missing
4. **Fix Missing Values** —
   - Text columns (`Category`, `Region`) were filled using the **mode** (most frequent value).
   - Numeric columns (`Sales`, `Quantity`) were filled using the **median** (middle value).
5. **Check for Duplicates** — No duplicate rows were found in the dataset.
6. **Save Cleaned Dataset** — Exported the cleaned data as `cleaned_retail_sales.csv`.

## Result
✅ All missing values were successfully fixed  
✅ No duplicate rows found  
✅ Final cleaned dataset saved and ready for further analysis

## Tools Used
- Python
- Pandas
- Kaggle Notebook

## Author
This project was built by https://github.com/AiWriter404 beginner learning Python and AI concepts.